### ISSUE TITLE
Getting inkscape-silhouette working with Windows

### ISSUE THREAD

**Issue by sctv** [fablabnbg/inkscape-silhouette#61]
Have tested the instructions and added to them. https://github.com/sctv/inkscape-silhouette/tree/patch-2
Have also proposed a work-around to get it working in Windows. https://github.com/sctv/inkscape-silhouette/tree/patch-3

**Comment by sctv** [comment]
Works for win7, failing on win10.

**Comment by sctv** [comment]
For WIn10:
Silhoutte on, usb recognised, paper loaded: 
Traceback (most recent call last):
  File "sendto_silhouette.py", line 1160, in <module>
    e.affect()
  File "inkex.py", line 283, in affect
    self.effect()
  File "sendto_silhouette.py", line 977, in effect
    state = dev.status()    # hint at loading paper, if not ready.
  File "silhouette/Graphtec.py", line 421, in status
    s.write("\x1b\x05")
  File "silhouette/Graphtec.py", line 356, in write
    raise ValueError('write %d bytes failed: r=%d' % (len(chunk), r))
ValueError: write 2 bytes failed: r=0

**Comment by jnweiger** [comment]
Hey Brian!

Thanks for the windows fix!
I've removed two typos in README.md, commited that directly to your patch-2
branch. Sorry for not makeing a proper PR.

cheers, W-

**Comment by sctv** [comment]
Hei @jnweiger

Thats fine.... my typos don't merit a pr. 
Happy to help. 

Have used your extension on raspberrypi over vnc/headerless. 
  
Still new to github. At the moment it works for win7. However, if you change the driver to work with inkscape it dosen't work with the manufacturers software. That software cant find the sihlouette portrait anymore. 

cheers, W-

**Comment by innocuo** [comment]
@sctv thanks man, your isntructions helped a lot. 
I made it work on win10. 

I was having the same error message as in the 3rd post in this thread

I run the following code before anything:

```
import usb.core

dev = usb.core.find(idVendor=0x0b4d, idProduct=0x1123)
dev.set_configuration()
cfg = dev.get_active_configuration()
intf = cfg[(0,0)]

ep = usb.util.find_descriptor(
    intf,
    # match the first OUT endpoint
    custom_match = \
    lambda e: \
        usb.util.endpoint_direction(e.bEndpointAddress) == \
        usb.util.ENDPOINT_OUT)
		
ep.write('test')
```


This is pretty much copy pasted from [pyusb's tutorial](https://github.com/pyusb/pyusb/blob/master/docs/tutorial.rst)
I got the vendorId and productId from Graphtec.py. I'm using the one for the Silhouette Portrait.

Basically it gets the device, and send a test message (I'm guessing "test" is not really the best option, but it works). When I run that script, I hear a quick sound from the Silhouette. After that using the Inkscape extension works ok.

So, my guess is Inkscape doesn't "wake up" the device, but once it's "woken" it works ok.

I also had another issue, not sure if related to this thread. Graphtec.py checks for pyusb's version, but it was thinking my version was "unknown". Apparently usb.version_info doesn't exist for my version of pyusb. So I just commented out that version check.

I hope this info helps

**Comment by sctv** [comment]
@innocuo thanks for that. Glad that my instructions worked, or gave you an idea for that.
I will try that. 
Had rewritten the usb_verion check for windows in Graftec.py in the work around. Commenting that is fine for Windows users, but  might be getting rid of useful info for linux troubleshooting.

Cheers,
B

**Comment by CombatCube** [comment]
I am having the same issue on Windows 10. The Silhouette blinks, but I get the write 2 bytes error, even after running the "initialization" code that @innocuo posted. Can you elaborate on where you put that code?

**Comment by sctv** [comment]
Have you rebooted the machine? It's a silly question...but mine started
working when I tried it again yesterday without any of the initialisation.
(On second thoughts, I may have updated inkscape- to the latest as well)

**Comment by innocuo** [comment]
@CombatCube inkscape-silhouette uses [pyusb](https://github.com/pyusb/pyusb) to connect and communicate with the Silhouette.

The code I posted before is the most basic code to use pyusb to connect to a usb device and send a test message. You can check out [pyusb's tutorial](https://github.com/pyusb/pyusb/blob/master/docs/tutorial.rst).

You run it independently from inkspace-silhouette because in reality it has nothing to do with it.
You save the code in a file (say "connect.py"), then run it with python in the command line ("python connect.py"). In my case, I used python that comes with inkscape. (I think it's in "C:\Program Files\Inkscape\python", I don't have the machine I tested on with me). You have to run it with a python installation that has pyusb installed

Also note that my code was intended for the Portrait 2. If you have any other device, you'd need to change the idProduct. I got the id from [Graphtec.py](https://github.com/fablabnbg/inkscape-silhouette/blob/master/silhouette/Graphtec.py), lines 118-140

If you run it, and either you don't get an error or better yet, the Silhouette reacts, it basically confirms that python can connect and communicate with the SIlhouette. And what I'm seeing is that once the Silhouette "wakes up", inkscape just works.

**Comment by sctv** [comment]
@CombatCube   did you get it to work?

**Comment by CombatCube** [comment]
Yes, just last night, by running that code standalone and then running the extension. Now to get it all working in the extension...

**Comment by sctv** [comment]
I dont need to use the standalone code after the first go. 
@CombatCube have you tried it on its own?

**Comment by EtherGraf** [comment]
I have no windows here, so i cannot test it. 

From looking at the code I can see the following lines in graphtec.py line 230:

      if sys_platform.startswith('win'):
        print("device init under windows not implemented. Help adding code!", file=self.log)

Maybe it helps to insert additional the code 

              dev.set_configuration()

This seems what is done in linux and in the example, but missing on windows.

Could anyone with windows test this please?

**Comment by jnweiger** [comment]
We have a new inkscape-silhouette-windows.zip bundle in the releases section. Maybe worth testing!
https://github.com/fablabnbg/inkscape-silhouette/releases/tag/v1.21.mose
